### PR TITLE
Async support

### PR DESCRIPTION
**Pull request by cwallsfdc**
Adds support for async requests based on `cloudevent.type`.  If async, middleware will forward based on headers and ignore the forward request's response.

```
# Initial async request...
2020-07-04T16:11:25 New invocation started
2020-07-04T16:11:25 Streaming pipeline initialized
2020-07-04T16:11:25 Ready to process data
2020-07-04T16:11:25 Received initial async request
2020-07-04T16:11:25 Forwarded async request

# Forwarded async request...
2020-07-04T16:11:25 New invocation started
2020-07-04T16:11:25 Streaming pipeline initialized
2020-07-04T16:11:25 Ready to process data
2020-07-04T16:11:25 Fulfilling async request
2020-07-04T16:11:25 DEBUG flag is unset
2020-07-04T16:11:25 Invoking AsyncApexFunction with payload {}
``

### CODE REVIEW

File: middleware/lib/FunctionInvocationRequest.ts [hunk 1]
```diff
@@ -0,0 +1,87 @@
+
```
Comment (cwallsfdc) [inline middleware/lib/FunctionInvocationRequest.ts:1]: Relocated to separate lib.

File: middleware/lib/sfMiddleware.ts [hunk 2]
```diff
@@ -0,0 +1,204 @@
+import {Logger, LoggerLevel} from '@salesforce/core/lib/logger';
```
Comment (cwallsfdc) [inline middleware/lib/sfMiddleware.ts:1]: Relocated to separate lib file.

File: middleware/index.ts [hunk 3]
```diff
@@ -282,6 +44,64 @@ function errorMessage(error: Error): any {
...
+    return new Promise((resolve, reject) => {
+        const expectedErrCode = 'EXPECTED_ECONNRESET';
+        const req = lib.request(options);
+        req.on('error', async (err) => {
+            if (req.destroyed && err['code'] === expectedErrCode) {
+                // Expected to terminate response
+                resolve();
+            } else {
+                // Return error for async request client
+                reject(err);
```
Comment (cwallsfdc) [inline middleware/index.ts:86]: Return errors encountered during the request back to original client to decide what to do.

File: middleware/index.ts [hunk 4]
```diff
@@ -293,12 +113,50 @@ export default async function systemFn(message: any): Promise<any> {
...
+    let isAsync = false;
     try {
-        const middlewareResult = await applySfFxMiddleware({payload, headers}, requestLogger);
-        const result = await userFn(...middlewareResult);
+        // If initial async request, invoke function again and release request
+        isAsync = isAsyncRequest(payload.type);
+        if (isAsync) {
+            if (!headers[ASYNC_FULFILL_HEADER]) {
+                requestLogger.info('Received initial async request');
+                await invokeAsyncFn(requestLogger, payload, headers);
```
Comment (jabrown85) [inline middleware/index.ts:123]: How does this end up working? The `202` response is after the promise is resolved in anyway, which doesn't seem to happen outside of `error`. 

If I understand correctly, we are sending the request body and then immediately forcing an error path to essentially "fire and forget" with connection error handling...is that right?

File: middleware/lib/FunctionInvocationRequest.ts [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+export enum FunctionInvocationRequestStatusEnum {
+  Success = 'Success',
+  Error = 'Error'
+}
+
+// Save update to FunctionInvocationRequest
+// Does not throw
+export async function saveFnInvocation(logger: Logger, 
+                                       fnInvocation: FunctionInvocationRequest, 
+                                       response: any,
```
Comment (jabrown85) [inline middleware/lib/FunctionInvocationRequest.ts:19]: I'm a bit concerned about `response` here. This could be a string, object, or even a promise. Today we only handle if it is falsely, but now we are trying to log the response back to core. We may need to test to see if it is a promise and handle that properly. As of right now, I think the `save` code below would essentially serialize the promise to a string.

File: middleware/package.json [hunk 6]
```diff
@@ -27,13 +27,14 @@
   "author": "TODO",
   "dependencies": {
     "@projectriff/message": "^1.0.0",
-    "@salesforce/salesforce-sdk": "^1.3.0"
+    "@salesforce/salesforce-sdk": "^1.3.0",
+    "tslib": "^1.11.2"
```
Comment (jabrown85) [inline middleware/package.json:31]: What made us move `tslib` from `devDeps` to `deps`?

File: middleware/test/unit/ContextTests.ts [hunk 7]
```diff
@@ -186,11 +185,7 @@ describe('Context Tests', () => {
         const fsStat = new fs.Stats();
         sinon.stub(fsStat, 'isDirectory').returns(true);
         sinon.stub(fsStat, 'isFile').returns(true);
-        sandbox.stub(fs, 'statSync')
-          .withArgs(`/platform/services/${sname}/secret`)
-          .returns(fsStat)
-          .withArgs(`/platform/services/${sname}/secret/${key}`)
-          .returns(fsStat);
+        sandbox.stub(fs, 'statSync').returns(fsStat);
```
Comment (jabrown85) [inline middleware/test/unit/ContextTests.ts:188]: Why the change here? Before the change any unexpected paths would make a test fail I think.

File: middleware/package.json (hunk 6 above)
Comment (amphro) [reply middleware/package.json:31]: Shouldn't tslib always be a dependency? It is a runtime library.

File: middleware/index.ts (hunk 4 above)
Comment (cwallsfdc) [reply middleware/index.ts:123]: If there's an error on the request, we'll return that to the client (Core) for it to decide what to do.

>>> is that right?
Yes.  `req.destory()` throws a custom error.

File: middleware/package.json (hunk 6 above)
Comment (jabrown85) [reply middleware/package.json:31]: I'm not sure. I guess maybe it worked before because of other things bringing it in? I'm 👍 , just like to call out dep location changes in case they were accidental.

File: middleware/lib/FunctionInvocationRequest.ts [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+export enum FunctionInvocationRequestStatusEnum {
+  Success = 'Success',
+  Error = 'Error'
+}
+
+// Save update to FunctionInvocationRequest
+// Does not throw
+export async function saveFnInvocation(logger: Logger,
+                                       fnInvocation: FunctionInvocationRequest,
+                                       response: any,
```
Comment (cwallsfdc) [inline middleware/lib/FunctionInvocationRequest.ts:19]: @jabrown85 comment was lost on the latest push.

Chatted on Slack.  In a nutshell, we'll monitor whether a developer returns an unresolved `Promise` w/in the returned userFn resolved `Promise`.
